### PR TITLE
Feature/add support for deprecate aws ssm maintenance window task

### LINKED ISSUE #5045: aws_ssm_maintenance_window_task missing task_invocation_parameters
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Unable to set invocation parameters on SSM maintenance task

### New or Affected Resource(s)

* aws_ssm_maintenance_window_task

### References

* [CloudFormation](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ssm-maintenancewindowtask.html#cfn-ssm-maintenancewindowtask-taskinvocationparameters)
* [AWS CLI](https://docs.aws.amazon.com/cli/latest/reference/ssm/register-task-with-maintenance-window.html)

### LINKED ISSUE #6431: extend aws_ssm_maintenance_window_task resource to accept name, description, task-parameters and Document version.
based on the cli options, https://docs.aws.amazon.com/cli/latest/reference/ssm/register-task-with-maintenance-window.html

You can set a name, description to the automation task.

Moreover, I am using task type AUTOMATION to start/stop instances. We need to be able to pass InstanceId via task parameters which currently the resource doesn't support.

### LINKED ISSUE #7154: aws_ssm_maintenance_window_task logging_info s3_bucket_name doesn't support update
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

Terraform v0.11.11
provider "aws" (1.56.0)

### Affected Resource(s)

* aws_ssm_maintenance_window_task

### Terraform Configuration Files

I've removed all non-relevant data:

Original state:
```hcl

resource "aws_ssm_maintenance_window_task" "foo" {
  window_id        = "${aws_ssm_maintenance_window.foo.id}"
  task_type        = "RUN_COMMAND"
  task_arn         = "AWS-RunPatchBaseline"
  service_role_arn = "${aws_iam_service_linked_role.ssm_maintenance_service_role.arn}"

  max_concurrency = "1"
  max_errors      = "100"

  targets {
    key    = "WindowTargetIds"
    values = ["${aws_ssm_maintenance_window_target.foo.id}"]
  }

  task_parameters {
    name   = "Operation"
    values = ["Install"]
  }

  logging_info {
    s3_bucket_name   = "foo"
    s3_bucket_prefix = "amazon-patch"
    s3_region        = "eu-west-1"
  }
}

```

Desired state:
```hcl

resource "aws_ssm_maintenance_window_task" "foo" {
  window_id        = "${aws_ssm_maintenance_window.foo.id}"
  task_type        = "RUN_COMMAND"
  task_arn         = "AWS-RunPatchBaseline"
  service_role_arn = "${aws_iam_service_linked_role.ssm_maintenance_service_role.arn}"

  max_concurrency = "1"
  max_errors      = "100"

  targets {
    key    = "WindowTargetIds"
    values = ["${aws_ssm_maintenance_window_target.foo.id}"]
  }

  task_parameters {
    name   = "Operation"
    values = ["Install"]
  }

  logging_info {
    s3_bucket_name   = "bar"
    s3_bucket_prefix = "amazon-patch"
    s3_region        = "eu-west-1"
  }
}

```

### Expected Behavior

Terraform updates the S3 bucket used for logging, either by updating the resource in-place or recreating the resource

### Actual Behavior

* aws_ssm_maintenance_window_task.foo: doesn't support update

### Steps to Reproduce

1. `terraform apply` the original state
2. `terraform apply` the desired state to update the logging bucket name

### LINKED ISSUE #7746: Feature Request - Add support for LAMBDA in aws_ssm_maintenance_window_task resource
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Similar to #4408 
I could do with support for registering a Lambda task with a maintenance window. In fact the current implementation uses 

```
TaskParameters={
        'string': {
            'Values': [
                'string',
            ]
        }
    },
```

Which has been deprecated in favour of 

```
    TaskInvocationParameters={
        'RunCommand': {
            'Comment': 'string',
            'DocumentHash': 'string',
            'DocumentHashType': 'Sha256'|'Sha1',
            'NotificationConfig': {
                'NotificationArn': 'string',
                'NotificationEvents': [
                    'All'|'InProgress'|'Success'|'TimedOut'|'Cancelled'|'Failed',
                ],
                'NotificationType': 'Command'|'Invocation'
            },
            'OutputS3BucketName': 'string',
            'OutputS3KeyPrefix': 'string',
            'Parameters': {
                'string': [
                    'string',
                ]
            },
            'ServiceRoleArn': 'string',
            'TimeoutSeconds': 123
        },
        'Automation': {
            'DocumentVersion': 'string',
            'Parameters': {
                'string': [
                    'string',
                ]
            }
        },
        'StepFunctions': {
            'Input': 'string',
            'Name': 'string'
        },
        'Lambda': {
            'ClientContext': 'string',
            'Qualifier': 'string',
            'Payload': b'bytes'
        }
    },
```

Implementing this would provide support for all 4 types of task

### New or Affected Resource(s)

* resource_aws_ssm_maintenance_window_task.go

### References

https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/ssm.html#SSM.Client.register_task_with_maintenance_window

### LINKED ISSUE #7870: aws provider does not appear to support --task-invocation-parameters for aws_ssm_maintenance_window_task resource
aws provider does not appear to support --task-invocation-parameters for aws_ssm_maintenance_window_task resource.  It does support the now deprecated --task-parameters which does not appear to support the needed complex json structures that the newer --task-invocation-parameters supports for automation docs.

### PR DESCRIPTION
**Pull request by kterada0509**
Closes #7870 
Closes #7746 
Closes #7154
Closes #6431
Closes #5045 

Reference #5700
Reference #4408
Reference #3218

Changes proposed in this pull request:

* [x] Fix to use `GetMaintenanceWindowTask` method when read resource
* [x] Add Support Update resource
* [x] Fix to not return error when resource is already deleted
* [x] Add support import exist resource
* [x] Add attribute `task_invocation_parameters`
* [x] Add deprecated message for `logging_info` and `task_parameters`
* [x] Add test case for `automation_parameters`
* [x] Add test case for `lambda_parameters`
* [x] Add test case for `run_command_parameters`
* [x] Add test case for `step_functions_parameters` 
* [x] Update document

Output from acceptance testing:

```
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSSSMMaintenanceWindowTask_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -parallel 20 -run=TestAccAWSSSMMaintenanceWindowTask_ -timeout 120m
=== RUN   TestAccAWSSSMMaintenanceWindowTask_basic
=== PAUSE TestAccAWSSSMMaintenanceWindowTask_basic
=== RUN   TestAccAWSSSMMaintenanceWindowTask_updateForcesNewResource
=== PAUSE TestAccAWSSSMMaintenanceWindowTask_updateForcesNewResource
=== RUN   TestAccAWSSSMMaintenanceWindowTask_TaskInvocationAutomationParameters
=== PAUSE TestAccAWSSSMMaintenanceWindowTask_TaskInvocationAutomationParameters
=== RUN   TestAccAWSSSMMaintenanceWindowTask_TaskInvocationLambdaParameters
=== PAUSE TestAccAWSSSMMaintenanceWindowTask_TaskInvocationLambdaParameters
=== RUN   TestAccAWSSSMMaintenanceWindowTask_TaskInvocationRunCommandParameters
=== PAUSE TestAccAWSSSMMaintenanceWindowTask_TaskInvocationRunCommandParameters
=== RUN   TestAccAWSSSMMaintenanceWindowTask_TaskInvocationStepFunctionParameters
=== PAUSE TestAccAWSSSMMaintenanceWindowTask_TaskInvocationStepFunctionParameters
=== CONT  TestAccAWSSSMMaintenanceWindowTask_basic
=== CONT  TestAccAWSSSMMaintenanceWindowTask_TaskInvocationRunCommandParameters
=== CONT  TestAccAWSSSMMaintenanceWindowTask_TaskInvocationStepFunctionParameters
=== CONT  TestAccAWSSSMMaintenanceWindowTask_TaskInvocationAutomationParameters
=== CONT  TestAccAWSSSMMaintenanceWindowTask_TaskInvocationLambdaParameters
=== CONT  TestAccAWSSSMMaintenanceWindowTask_updateForcesNewResource
--- PASS: TestAccAWSSSMMaintenanceWindowTask_TaskInvocationLambdaParameters (147.07s)
--- PASS: TestAccAWSSSMMaintenanceWindowTask_TaskInvocationStepFunctionParameters (163.38s)
--- PASS: TestAccAWSSSMMaintenanceWindowTask_TaskInvocationAutomationParameters (169.52s)
--- PASS: TestAccAWSSSMMaintenanceWindowTask_basic (173.94s)
--- PASS: TestAccAWSSSMMaintenanceWindowTask_TaskInvocationRunCommandParameters (204.06s)
--- PASS: TestAccAWSSSMMaintenanceWindowTask_updateForcesNewResource (232.43s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	232.506s
```

### CODE REVIEW

**Comment by e-moshaya** [comment]
any update on this PR when will this get merged?

**Comment by AndrewCi** [comment]
Will this address https://github.com/terraform-providers/terraform-provider-aws/issues/2876?

**Comment by kterada0509** [comment]
Re-run acctest
```
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSSSMMaintenanceWindowTask_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -parallel 20 -run=TestAccAWSSSMMaintenanceWindowTask_ -timeout 120m
=== RUN   TestAccAWSSSMMaintenanceWindowTask_basic
=== PAUSE TestAccAWSSSMMaintenanceWindowTask_basic
=== RUN   TestAccAWSSSMMaintenanceWindowTask_updateForcesNewResource
=== PAUSE TestAccAWSSSMMaintenanceWindowTask_updateForcesNewResource
=== RUN   TestAccAWSSSMMaintenanceWindowTask_TaskInvocationAutomationParameters
=== PAUSE TestAccAWSSSMMaintenanceWindowTask_TaskInvocationAutomationParameters
=== RUN   TestAccAWSSSMMaintenanceWindowTask_TaskInvocationLambdaParameters
=== PAUSE TestAccAWSSSMMaintenanceWindowTask_TaskInvocationLambdaParameters
=== RUN   TestAccAWSSSMMaintenanceWindowTask_TaskInvocationRunCommandParameters
=== PAUSE TestAccAWSSSMMaintenanceWindowTask_TaskInvocationRunCommandParameters
=== RUN   TestAccAWSSSMMaintenanceWindowTask_TaskInvocationStepFunctionParameters
=== PAUSE TestAccAWSSSMMaintenanceWindowTask_TaskInvocationStepFunctionParameters
=== CONT  TestAccAWSSSMMaintenanceWindowTask_basic
=== CONT  TestAccAWSSSMMaintenanceWindowTask_TaskInvocationStepFunctionParameters
=== CONT  TestAccAWSSSMMaintenanceWindowTask_TaskInvocationRunCommandParameters
=== CONT  TestAccAWSSSMMaintenanceWindowTask_TaskInvocationLambdaParameters
=== CONT  TestAccAWSSSMMaintenanceWindowTask_TaskInvocationAutomationParameters
=== CONT  TestAccAWSSSMMaintenanceWindowTask_updateForcesNewResource
--- PASS: TestAccAWSSSMMaintenanceWindowTask_TaskInvocationLambdaParameters (148.07s)
--- PASS: TestAccAWSSSMMaintenanceWindowTask_TaskInvocationStepFunctionParameters (151.84s)
--- PASS: TestAccAWSSSMMaintenanceWindowTask_TaskInvocationAutomationParameters (183.01s)
--- PASS: TestAccAWSSSMMaintenanceWindowTask_updateForcesNewResource (188.49s)
--- PASS: TestAccAWSSSMMaintenanceWindowTask_TaskInvocationRunCommandParameters (194.87s)
--- PASS: TestAccAWSSSMMaintenanceWindowTask_basic (214.83s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	214.939s
```

File: aws/resource_aws_ssm_maintenance_window_task_test.go [hunk 1]
```diff
@@ -274,3 +433,517 @@ EOF
...
+  service_role_arn = "${aws_iam_role.ssm_role.arn}"
+  max_concurrency = "2"
+  max_errors = "1"
+  targets {
+    key = "InstanceIds"
+    values = ["${aws_instance.foo.id}"]
+  }
+  task_invocation_parameters {
+		run_command_parameters {
+			comment 							= "%[2]s"
```
[the remark this answers is not in the document]
Comment (kterada0509) [reply aws/resource_aws_ssm_maintenance_window_task_test.go]: I fixed.

File: website/docs/r/ssm_maintenance_window_task.html.markdown [hunk 2]
```diff
@@ -63,8 +63,9 @@ The following arguments are supported:
 * `description` - (Optional) The description of the maintenance window task.
 * `targets` - (Required) The targets (either instances or window target ids). Instances are specified using Key=InstanceIds,Values=instanceid1,instanceid2. Window target ids are specified using Key=WindowTargetIds,Values=window target id1, window target id2.
 * `priority` - (Optional) The priority of the task in the Maintenance Window, the lower the number the higher the priority. Tasks in a Maintenance Window are scheduled in priority order with tasks that have the same priority scheduled in parallel.
-* `logging_info` - (Optional) A structure containing information about an Amazon S3 bucket to write instance-level logs to. Documented below.
-* `task_parameters` - (Optional) A structure containing information about parameters required by the particular `task_arn`. Documented below.
+* `logging_info` - (Optional,**Deprecated**) A structure containing information about an Amazon S3 bucket to write instance-level logs to. Documented below.
```
[the remark this answers is not in the document]
Comment (kterada0509) [reply website/docs/r/ssm_maintenance_window_task.html.markdown]: I fixed.

File: website/docs/r/ssm_maintenance_window_task.html.markdown [hunk 3]
```diff
@@ -77,8 +78,64 @@ The following arguments are supported:
 * `name` - (Required)
 * `values` - (Required)
 
+`task_invocation_parameters` supports the following:
+
+This argument is conflict with `task_parameters` and `logging_info`.
```
[the remark this answers is not in the document]
Comment (kterada0509) [reply website/docs/r/ssm_maintenance_window_task.html.markdown]: I fixed.

File: website/docs/r/ssm_maintenance_window_task.html.markdown [hunk 4]
```diff
@@ -77,8 +78,64 @@ The following arguments are supported:
...
+
+* `automation_parameters` - (Optional) The parameters for an AUTOMATION task type. Documented below.
+* `lambda_parameters` - (Optional) The parameters for a LAMBDA task type. Documented below.
+* `run_command_parameters` - (Optional) The parameters for a RUN_COMMAND task type. Documented below.
+* `step_functions_parameters` - (Optional)
```
[the remark this answers is not in the document]
Comment (kterada0509) [reply website/docs/r/ssm_maintenance_window_task.html.markdown]: I fixed.

File: website/docs/r/ssm_maintenance_window_task.html.markdown [hunk 5]
```diff
@@ -63,8 +63,9 @@ The following arguments are supported:
 * `description` - (Optional) The description of the maintenance window task.
 * `targets` - (Required) The targets (either instances or window target ids). Instances are specified using Key=InstanceIds,Values=instanceid1,instanceid2. Window target ids are specified using Key=WindowTargetIds,Values=window target id1, window target id2.
 * `priority` - (Optional) The priority of the task in the Maintenance Window, the lower the number the higher the priority. Tasks in a Maintenance Window are scheduled in priority order with tasks that have the same priority scheduled in parallel.
-* `logging_info` - (Optional) A structure containing information about an Amazon S3 bucket to write instance-level logs to. Documented below.
-* `task_parameters` - (Optional) A structure containing information about parameters required by the particular `task_arn`. Documented below.
+* `logging_info` - (Optional,**Deprecated**) A structure containing information about an Amazon S3 bucket to write instance-level logs to. Documented below.
+* `task_parameters` - (Optional,**Deprecated**) A structure containing information about parameters required by the particular `task_arn`. Documented below.
```
[the remark this answers is not in the document]
Comment (kterada0509) [reply website/docs/r/ssm_maintenance_window_task.html.markdown]: I fixed.

File: aws/resource_aws_ssm_maintenance_window_task.go [hunk 6]
```diff
@@ -45,30 +50,25 @@ func resourceAwsSsmMaintenanceWindowTask() *schema.Resource {
...
 			"service_role_arn": {
 				Type:     schema.TypeString,
 				Required: true,
-				ForceNew: true,
 			},
 
 			"targets": {
 				Type:     schema.TypeList,
 				Required: true,
-				ForceNew: true,
```
[the remark this answers is not in the document]
Comment (kterada0509) [reply aws/resource_aws_ssm_maintenance_window_task.go]: I added test case.

File: aws/resource_aws_ssm_maintenance_window_task.go [hunk 7]
```diff
@@ -45,30 +50,25 @@ func resourceAwsSsmMaintenanceWindowTask() *schema.Resource {
 			"task_arn": {
 				Type:     schema.TypeString,
 				Required: true,
-				ForceNew: true,
 			},
 
 			"service_role_arn": {
 				Type:     schema.TypeString,
 				Required: true,
-				ForceNew: true,
```
[the remark this answers is not in the document]
Comment (kterada0509) [reply aws/resource_aws_ssm_maintenance_window_task.go]: I added test case.

File: aws/resource_aws_ssm_maintenance_window_task.go [hunk 8]
```diff
@@ -45,30 +50,25 @@ func resourceAwsSsmMaintenanceWindowTask() *schema.Resource {
 			"task_arn": {
 				Type:     schema.TypeString,
 				Required: true,
-				ForceNew: true,
```
[the remark this answers is not in the document]
Comment (kterada0509) [reply aws/resource_aws_ssm_maintenance_window_task.go]: I added test case.

File: aws/resource_aws_ssm_maintenance_window_task.go [hunk 9]
```diff
@@ -27,13 +34,11 @@ func resourceAwsSsmMaintenanceWindowTask() *schema.Resource {
 			"max_concurrency": {
 				Type:     schema.TypeString,
 				Required: true,
-				ForceNew: true,
 			},
 
 			"max_errors": {
 				Type:     schema.TypeString,
 				Required: true,
-				ForceNew: true,
```
[the remark this answers is not in the document]
Comment (kterada0509) [reply aws/resource_aws_ssm_maintenance_window_task.go]: I added test case.

File: aws/resource_aws_ssm_maintenance_window_task.go [hunk 10]
```diff
@@ -27,13 +34,11 @@ func resourceAwsSsmMaintenanceWindowTask() *schema.Resource {
 			"max_concurrency": {
 				Type:     schema.TypeString,
 				Required: true,
-				ForceNew: true,
```
[the remark this answers is not in the document]
Comment (kterada0509) [reply aws/resource_aws_ssm_maintenance_window_task.go]: I added test case.

**Comment by bflad** [comment]
Hi again @kterada0509 👋 Thank you for those updates. Did you see the feedback that the GitHub user interface hid behind the "19 hidden conversations" link?

Screen Shot 2019-07-12 at 3 39 10 PM

It looks like some or many of them are still pending as well. You can also see all review feedback by clicking the link there or going to the Files tab of the pull request. Please let us know. Thanks.

**Comment by kterada0509** [comment]
Re-run acctest.
```
ssm_maintenance_window_task *)$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSSSMMaintenanceWindowTask_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -parallel 20 -run=TestAccAWSSSMMaintenanceWindowTask_ -timeout 120m
=== RUN   TestAccAWSSSMMaintenanceWindowTask_basic
=== PAUSE TestAccAWSSSMMaintenanceWindowTask_basic
=== RUN   TestAccAWSSSMMaintenanceWindowTask_updateForcesNewResource
=== PAUSE TestAccAWSSSMMaintenanceWindowTask_updateForcesNewResource
=== RUN   TestAccAWSSSMMaintenanceWindowTask_TaskInvocationAutomationParameters
=== PAUSE TestAccAWSSSMMaintenanceWindowTask_TaskInvocationAutomationParameters
=== RUN   TestAccAWSSSMMaintenanceWindowTask_TaskInvocationLambdaParameters
=== PAUSE TestAccAWSSSMMaintenanceWindowTask_TaskInvocationLambdaParameters
=== RUN   TestAccAWSSSMMaintenanceWindowTask_TaskInvocationRunCommandParameters
=== PAUSE TestAccAWSSSMMaintenanceWindowTask_TaskInvocationRunCommandParameters
=== RUN   TestAccAWSSSMMaintenanceWindowTask_TaskInvocationStepFunctionParameters
=== PAUSE TestAccAWSSSMMaintenanceWindowTask_TaskInvocationStepFunctionParameters
=== CONT  TestAccAWSSSMMaintenanceWindowTask_basic
=== CONT  TestAccAWSSSMMaintenanceWindowTask_TaskInvocationRunCommandParameters
=== CONT  TestAccAWSSSMMaintenanceWindowTask_TaskInvocationStepFunctionParameters
=== CONT  TestAccAWSSSMMaintenanceWindowTask_TaskInvocationAutomationParameters
=== CONT  TestAccAWSSSMMaintenanceWindowTask_TaskInvocationLambdaParameters
=== CONT  TestAccAWSSSMMaintenanceWindowTask_updateForcesNewResource
--- PASS: TestAccAWSSSMMaintenanceWindowTask_TaskInvocationStepFunctionParameters (163.76s)
--- PASS: TestAccAWSSSMMaintenanceWindowTask_TaskInvocationLambdaParameters (186.35s)
--- PASS: TestAccAWSSSMMaintenanceWindowTask_updateForcesNewResource (196.74s)
--- PASS: TestAccAWSSSMMaintenanceWindowTask_basic (210.43s)
--- PASS: TestAccAWSSSMMaintenanceWindowTask_TaskInvocationAutomationParameters (229.00s)
--- PASS: TestAccAWSSSMMaintenanceWindowTask_TaskInvocationRunCommandParameters (234.95s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	235.059s
```

File: aws/resource_aws_ssm_maintenance_window_task.go [hunk 11]
```diff
@@ -78,28 +78,25 @@ func resourceAwsSsmMaintenanceWindowTask() *schema.Resource {
 			"name": {
 				Type:         schema.TypeString,
 				Optional:     true,
-				ForceNew:     true,
```
[the remark this answers is not in the document]
Comment (kterada0509) [reply aws/resource_aws_ssm_maintenance_window_task.go]: I added.

File: aws/resource_aws_ssm_maintenance_window_task.go [hunk 12]
```diff
@@ -78,28 +78,25 @@ func resourceAwsSsmMaintenanceWindowTask() *schema.Resource {
...
 				ValidateFunc: validateAwsSSMMaintenanceWindowTaskName,
 			},
 
 			"description": {
 				Type:         schema.TypeString,
 				Optional:     true,
-				ForceNew:     true,
```
[the remark this answers is not in the document]
Comment (kterada0509) [reply aws/resource_aws_ssm_maintenance_window_task.go]: I added.

File: aws/resource_aws_ssm_maintenance_window_task.go [hunk 13]
```diff
@@ -78,28 +78,25 @@ func resourceAwsSsmMaintenanceWindowTask() *schema.Resource {
...
 				ValidateFunc: validation.StringLenBetween(1, 128),
 			},
 
 			"priority": {
 				Type:     schema.TypeInt,
 				Optional: true,
-				ForceNew: true,
```
[the remark this answers is not in the document]
Comment (kterada0509) [reply aws/resource_aws_ssm_maintenance_window_task.go]: I added.

[the remark this answers is not in the document]
Comment (kterada0509) [reply aws/resource_aws_ssm_maintenance_window_task.go]: I added.

File: aws/resource_aws_ssm_maintenance_window_task.go [hunk 14]
```diff
@@ -78,28 +78,25 @@ func resourceAwsSsmMaintenanceWindowTask() *schema.Resource {
...
 
 			"logging_info": {
-				Type:     schema.TypeList,
-				MaxItems: 1,
-				Optional: true,
-				ForceNew: true,
+				Type:       schema.TypeList,
+				MaxItems:   1,
+				Optional:   true,
+				Deprecated: "use 'task_invocation_parameters' argument instead",
```
[the remark this answers is not in the document]
Comment (kterada0509) [reply aws/resource_aws_ssm_maintenance_window_task.go]: I added.

File: aws/resource_aws_ssm_maintenance_window_task.go [hunk 15]
```diff
@@ -119,25 +116,206 @@ func resourceAwsSsmMaintenanceWindowTask() *schema.Resource {
 			},
 
 			"task_parameters": {
-				Type:     schema.TypeList,
-				Optional: true,
-				ForceNew: true,
+				Type:       schema.TypeList,
+				Optional:   true,
+				Deprecated: "use 'task_invocation_parameters' argument instead",
```
[the remark this answers is not in the document]
Comment (kterada0509) [reply aws/resource_aws_ssm_maintenance_window_task.go]: I added.

File: aws/resource_aws_ssm_maintenance_window_task.go [hunk 16]
```diff
@@ -119,25 +116,206 @@ func resourceAwsSsmMaintenanceWindowTask() *schema.Resource {
...
+							Optional: true,
+							Elem: &schema.Resource{
+								Schema: map[string]*schema.Schema{
+									"document_version": {
+										Type:         schema.TypeString,
+										Optional:     true,
+										ValidateFunc: validation.StringMatch(regexp.MustCompile("([$]LATEST|[$]DEFAULT|^[1-9][0-9]*$)"), "see https://docs.aws.amazon.com/systems-manager/latest/APIReference/API_MaintenanceWindowAutomationParameters.html"),
+									},
+									"parameters": {
+										Type:     schema.TypeList,
```
[the remark this answers is not in the document]
Comment (kterada0509) [reply aws/resource_aws_ssm_maintenance_window_task.go]: fixed it.

File: aws/resource_aws_ssm_maintenance_window_task.go [hunk 17]
```diff
@@ -119,25 +116,206 @@ func resourceAwsSsmMaintenanceWindowTask() *schema.Resource {
...
+										Optional: true,
+									},
+
+									"output_s3_key_prefix": {
+										Type:     schema.TypeString,
+										Optional: true,
+									},
+
+									"parameters": {
+										Type:     schema.TypeList,
```
[the remark this answers is not in the document]
Comment (kterada0509) [reply aws/resource_aws_ssm_maintenance_window_task.go]: fixed it.

File: aws/resource_aws_ssm_maintenance_window_task.go [hunk 18]
```diff
@@ -195,6 +373,264 @@ func flattenAwsSsmTaskParameters(taskParameters map[string]*ssm.MaintenanceWindo
 	return result
 }
 
+func expandAwsSsmTaskInvocationParameters(config []interface{}) *ssm.MaintenanceWindowTaskInvocationParameters {
+	params := &ssm.MaintenanceWindowTaskInvocationParameters{}
+	for _, v := range config {
+		paramConfig := v.(map[string]interface{})
+		if attr, ok := paramConfig["automation_parameters"]; ok && len(attr.([]interface{})) > 0 {
```
[the remark this answers is not in the document]
Comment (kterada0509) [reply aws/resource_aws_ssm_maintenance_window_task.go]: fixed it.

File: aws/resource_aws_ssm_maintenance_window_task.go [hunk 19]
```diff
@@ -195,6 +373,264 @@ func flattenAwsSsmTaskParameters(taskParameters map[string]*ssm.MaintenanceWindo
...
+			params.Automation = expandAwsSsmTaskInvocationAutomationParameters(attr.([]interface{}))
+		}
+		if attr, ok := paramConfig["lambda_parameters"]; ok && len(attr.([]interface{})) > 0 {
```
[the remark this answers is not in the document]
Comment (kterada0509) [reply aws/resource_aws_ssm_maintenance_window_task.go]: fixed it.

File: aws/resource_aws_ssm_maintenance_window_task.go [hunk 20]
```diff
@@ -195,6 +373,264 @@ func flattenAwsSsmTaskParameters(taskParameters map[string]*ssm.MaintenanceWindo
...
+			params.Lambda = expandAwsSsmTaskInvocationLambdaParameters(attr.([]interface{}))
+		}
+		if attr, ok := paramConfig["run_command_parameters"]; ok && len(attr.([]interface{})) > 0 {
```
[the remark this answers is not in the document]
Comment (kterada0509) [reply aws/resource_aws_ssm_maintenance_window_task.go]: fixed it.

File: aws/resource_aws_ssm_maintenance_window_task.go [hunk 21]
```diff
@@ -195,6 +373,264 @@ func flattenAwsSsmTaskParameters(taskParameters map[string]*ssm.MaintenanceWindo
...
+			params.RunCommand = expandAwsSsmTaskInvocationRunCommandParameters(attr.([]interface{}))
+		}
+		if attr, ok := paramConfig["step_functions_parameters"]; ok && len(attr.([]interface{})) > 0 {
```
[the remark this answers is not in the document]
Comment (kterada0509) [reply aws/resource_aws_ssm_maintenance_window_task.go]: fixed it.

File: aws/resource_aws_ssm_maintenance_window_task_test.go [hunk 22]
```diff
@@ -274,3 +433,517 @@ EOF
...
+  }
+}
+
+resource "aws_instance" "foo" {
+  ami = "ami-4fccb37f"
+
+  instance_type = "m1.small"
+}
+
+resource "aws_iam_role" "ssm_role" {
```
[the remark this answers is not in the document]
Comment (kterada0509) [reply aws/resource_aws_ssm_maintenance_window_task_test.go]: fixed it.

**Pull request merged**